### ISSUE TITLE
Image Registration for hyperspectral data

### ISSUE THREAD

**Issue by amtamasi** [viscenter/registration-toolkit#34]
Right now, we can only perform image registration with 2D images. We would like to be able to perform registration with hyperspectral data cubes as well. A user should be able to register two images together, an image to a data cube, or two data cubes together. The output should be an image or a data cube, depending on the input.

i.e. register2D should be able to compute the transform between two objects of some Data (e.g. image or data cube), and apply it to the moving object.

**Comment by amtamasi** [comment]
Should just have to edit the functions/programs for computing the transform and applying the transform to act based on the type of input given (e.g. image, data cube, mesh). Some points to act on:

1. Modify apply_transform to take inputs as images or data cubes (or meshes). It should already account for meshes because it works with register3D.cpp.
 
2. Computing transformation is done in register2D.cpp and register3D.cpp. Should be able to modify these to take in images and data cubes. register3D should be able to take in a mesh as the 'fixed image', but that is already implemented.

**Comment by amtamasi** [comment]
Modifications to be made:
- apply_transform:
  change cv::imread so that we can be data type agnostic. We still need to return a cv Mat object regardless of data type, so we want to use cv::imread (somehow, when reading one image channel from data cube?) internal to whatever Data abstraction we write.
- register2D:
  Same as in apply_transform. Just need to fix how images are read in based on data type
- register3D:
  Fix how moving image is read in. Same concept as the previous two. Fixed image is a mesh

**Comment by amtamasi** [comment]
Idea for Data class abstraction. Subclasses: Image, Cube (split up into different cube file types like ENVI, FITS, etc.)
```
auto cvFixed = Data::readOneImage(fixedPath.string())
auto cvMoving = Data::readOneImage(movingPath.string()) 
```
OR
```
//Initialize as Data object, which decides what type this file is.
//Data* fixed is a pointer so that we can initialize subclass of the specific data type 
//and getOneImage() that way
//cvFixed and cvMoving are cv Mat objects

Data* fixed = Data(fixedPath.string())
auto cvFixed = fixed.getOneImage()

Data* moving = Data(movingPath.string())
auto cvMoving = moving.getOneImage()
```

**Comment by amtamasi** [comment]
In the code, you need to know the type of the fixed and moving inputs. It has to be more specific than the Data parent class, because the different types of data need to "getOneImage" in their specific way. 

For now, will plan to just set command line arguments like "--fixed-input-type envi" and have the default values be 'image'. In the future, can use metadata in the Registration package to know what kind of data it is.

**Comment by csparker247** [comment]
```C++
// Classes.hpp
class Data {
public:
    static Data* Load(std::string path);
    virtual cv::Mat getImage(int idx) = 0;
protected:
    Data() = default
}

class Image : public Data {
public:
    Image(std::string path) { /* do something to load*/ }
    cv::Mat getImage(int /* idx */ = 0) { return img_; }
private:
    cv::Mat img_; 
}

class ENVI : public Data {
public:
    ENVI(std::string path) { /* do something to load*/ }
    cv::Mat getImage(int idx) { return bands_[idx]; }
private:
   std::vector<cv::Mat> bands_;
}

// Classes.cpp
Data* Data::Load(std::string path) {
    json metadata(path);
    switch(metadata.type()) {
        case Type::SingleImage:
            return new Image(path);
        case Type::ENVI:
            return new ENVI(path);     
     }
}

// main.cpp
int main() {
    Data* image = Data::Load(imgPath);
    Data* cube = Data::Load(enviPath);

    cv::Mat fixed = image->getImage();
    cv::Mat moving = cube->getImage(10);   
}
```

This is the Factory design pattern. We don't really care which type of Data object it is as long as it has a common interface. All of this loading logic would be abstracted away inside of the regpkg so that regpkg gives you Data objects (though Data is becoming an increasingly generic name).

**Comment by amtamasi** [comment]
We need a better name for this 'Data' class we are moving towards. Data is too generic as an umbrella term for our input data. For our purpose, this class accommodates the following input data: images, meshes, hyper-spectral images (can be thought of as a 'stack' of images).

Any ideas are appreciated.

**Comment by csparker247** [comment]
> In register3D, the fixedImage is a 3D mesh. Should this be represented by the 'Data' object as well?

Yes, `Data` can potentially provide a textured mesh (mesh, image, and a UV map (2D<->3D mapping between the mesh and the texture image))

**Comment by amtamasi** [comment]
How should the metadata types be defined? Just an enum like:
```
//Data.cpp, Data::Load(std::string path) function
enum MetadataType { "image", "envi" };
MetadataType my_file_type = readFileType(path); //Read the file type from associated metadata.json file

switch(my_file_type){
    case "image":
        return new Image(path);
    case "envi":
        return new ENVI(path);
}
```
If this is the case, would that enum just be defined at the beginning of the Data::Load(path) function to load an object based on file type? Would we also use the "new" keyword, or instead make it a "unique_ptr<>"?

**Comment by amtamasi** [comment]
For register3D, can the fixed image always be assumed to be a mesh? I would like to just do a Data::Data::Load() call, but it looks like the OBJReader takes in a boost file path as a parameter, as opposed the the boostpath.string() that is usually passed in to Load(). 

If the fixed object can only be a mesh, we can just load it in via Data::Mesh::Load() and pass in the boost file path like the OBJReader wants to use.

**Comment by amtamasi** [comment]
OBJReader takes in a boost filesystem path, rather than the boostfilepath.string() I have been using. Should I change the general Data::Data::Load function to take in a boost path instead, and then call the .string() member function later if needed? Just need the actual boostfilepath for reading Mesh objects

**Comment by amtamasi** [comment]
A few thoughts: 
- Will likely need to include the envitools project in registration-toolkit so that we can deal with ENVI files.
- May want to adjust the ENVI class so that the constructor doesn't read in ALL the bands. Could just set up the ENVI file to be read from. Then, ENVI::getImage() member function could get load in a specified band as a cv::Mat
- How can a band be read in as a cv::Mat? Can it just be cast/declared that way?

**Comment by csparker247** [comment]
Notes:
- Image types should be an enum, and yes, the factory Load method will just switch the parsed type to load the actual image data. Use `enum class` instead of just `enum.`
- Use CMake's FetchContent if you want to include one of our other libraries in this project: https://cmake.org/cmake/help/latest/module/FetchContent.html. Look in the qt-refactor branch for Voluminous for an example of how this works.
- The ENVI class doesn't read all of the bands at construction. It only reads the metadata. It loads bands from disk when you call the `getBand(int)` function.
- `getBand` returns a cv::Mat.

**Comment by amtamasi** [comment]
Does envitools::envi::closeFile() close the the file reading from the "setAccessMode" member? Or does it close the file all together, needing to be read in again like 
`et::ENVI envi_(path)`?

**Comment by amtamasi** [comment]
Just some comments for me to remember that need to be implemented:
- Finish Data::Load function
- Rename the Data class/namespace and apply that new name
- Add envitools to registration-toolkit dependencies
  - Think this is in cmake/FindDependencies.cmake file of this project, but not 100% sure how to add in. Have essentially copied how another library was copied into volume-cartographer, but not sure I have the right pieces.

**Comment by csparker247** [comment]
> Does envitools::envi::closeFile() close the the file reading set up by the "setAccessMode" member? Or does it close the file all together, needing to be read in again like
> `et::ENVI envi_(path)`?
> 
> It looks like it closes the file completely, but I don't think we would want to do that, right? I plan to add a function to close the envi_ object (a private member), but it seems like we would want to keep that open until a user explicitly didn't want it anymore and closed it, trying to think ahead to using this through the GUI

`AccessMode::KeepOpen` simply means that we keep an open file handle to the ENVI file on-disk. If you're planning on frequently reading lots of bands at one time, this saves you the time it takes for the operating system to open a file descriptor every time you request a band. If you `closeFile(),` a new file descriptor will be opened the next time that `getBand()` is called. You don't need to reconstruct/reinitialize the ENVI object itself. This is purely a performance consideration, nothing functionally different.
